app/src/test/java/org/mozilla/fenix/components/metrics/AppStartupTelemetryTest.kt

```diff
@@ -0,0 +1,241 @@
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.HOT
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.ERROR
+import java.util.HashSet
+
+class AppStartupTelemetryTest {
+
+    @MockK
+    private lateinit var metricController: MetricController
+    @MockK
+    private lateinit var intent: Intent
+
+    private lateinit var appStartupTelemetry: AppStartupTelemetry
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+        appStartupTelemetry = AppStartupTelemetry(metricController)
+        every { metricController.track(any()) } returns Unit
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through application icon THEN records the correct values`() {
+        setupIntentMock(APP_ICON)
+
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(APP_ICON, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through a url link THEN records the correct values`() {
+        setupIntentMock(LINK)
+
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(LINK, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through an custom tab THEN records the correct values`() {
+        setupIntentMock(CUSTOM_TAB)
+
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onExternalAppBrowserOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(CUSTOM_TAB, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN, application  exists and is backgrounded, WHEN application is launched again through app icon and HomeActivity is recreated THEN records the correct values`() {
+        setupIntentMock(APP_ICON)
+        launchApplicationAndPutApplicationInBackground()
+
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppAllStartup(APP_ICON, WARM, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN, application  exists and is backgrounded, WHEN application is launched again through url link and HomeActivity is recreated THEN records the correct values`() {
```

Comment:
No action needed

Note: I notice you're writing a test for many combinations: 3 startup types foreground, 3 startup types background, etc.

I probably would have written all these tests 😅 but it probably is a perfect is the enemy of the good situation where it's probably not necessary: doing all the combinations in one type of test and hitting all the other types of tests is probably good enough.

That being said, junit supports a feature called parameterized tests which would make writing all these tests combinations much less code but I haven't ever been able to set it up.